contracts/VariableUtilisationRatesCalculator.sol

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.2;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./IRatesCalculator.sol";
+import "./WadRayMath.sol";
+import "hardhat/console.sol";
+
+/**
+ * @title VariableUtilisationRatesCalculator
+ * @dev Contract which calculates the interest rates based on pool utilisation.
+ * Utilisation is computed as the ratio between funds borrowed and funds deposited to the pool.
+ * Borrowing rates are calculated using a piecewise linear function. The first piece is defined by SLOPE_1
+ * and OFFSET (shift). Second piece is defined by SLOPE_2 (calculated off-chain), BREAKPOINT (threshold value above
+ * which second piece is considered) and MAX_RATE (value at pool utilisation of 1).
+**/
+contract VariableUtilisationRatesCalculator is IRatesCalculator, Ownable {
+  using WadRayMath for uint256;
+
+  uint256 public constant SLOPE_1 = 0.12 ether;
+  uint256 public constant OFFSET = 0.05 ether;
+  // BREAKPOINT must be lower than 1 ether
+  uint256 public constant BREAKPOINT = 0.8 ether;
+  uint256 public constant MAX_RATE = 0.75 ether;
+
+  // calculated off-chain for gas efficiency with following formula:
+  // (MAX_RATE - OFFSET - SLOPE_1 * BREAKPOINT) / (1 - BREAKPOINT)
+  uint256 public constant SLOPE_2 = 3.02 ether;
+
+
+/* ========== VIEW FUNCTIONS ========== */
+
+
+  /**
+    * Returns the pool utilisation, which is a ratio between loans and deposits
+    * utilisation = value_of_loans / value_of_deposits
+    * @dev _totalLoans total value of loans
+    * @dev _totalDeposits total value of deposits
+  **/
+  function getPoolUtilisation(uint256 _totalLoans, uint256 _totalDeposits) public pure returns (uint256) {
+    if (_totalDeposits == 0) return 0;
+
+    return _totalLoans.wadToRay()
+    .rayDiv(_totalDeposits.wadToRay())
+    .rayToWad();
+  }
+
+
+  /**
+    * Returns the current deposit rate
+    * The value is based on the current borrowing rate and satisfies the invariant:
+    * value_of_loans * borrowing_rate = value_of_deposits * deposit_rate
+    * @dev _totalLoans total value of loans
+    * @dev _totalDeposits total value of deposits
+  **/
+  function calculateDepositRate(uint256 _totalLoans, uint256 _totalDeposits) external view override returns (uint256) {
+    if (_totalDeposits == 0) return 0;
+
+    return this.calculateBorrowingRate(_totalLoans, _totalDeposits).wadToRay()
```

Comment:
what if totalLoans > totalDeposits
it seems like the depositRate > borrowingRate?

How about capping the utilisation at 100% max

Reply:
this scenario is resolved https://github.com/ava-loan/avaloan/pull/17. Summing up calculated poolUtilisation can be higher than 1, but in this case `calculateDepositRate` and  `calculateBorrowingRate` return maximum possible values `MAX_RATE`

Reply:
https://github.com/ava-loan/avaloan/blob/3857bd9a2be2d7c8a8fc8c1b18aafd75bbd9aabb/contracts/VariableUtilisationRatesCalculator.sol#L58

https://github.com/ava-loan/avaloan/blob/3857bd9a2be2d7c8a8fc8c1b18aafd75bbd9aabb/contracts/VariableUtilisationRatesCalculator.sol#L84

-> btw probably here is better to put <=

Reply:
changed to >=